File: src/bin/analyze_albums/name_utils.rs
use std::collections::HashSet;

pub(super) fn normalize_name(v: &str) -> String {
    fold_fullwidth_ascii(v)
        .to_lowercase()
        .chars()
        .filter(|c| !c.is_whitespace())
        .collect::<String>()
        .replace(['\'', '“', '”'], "\"")
}

pub(super) fn fold_fullwidth_ascii(input: &str) -> String {
    input
        .chars()
        .map(|c| match c {
            '\u{3000}' => ' ',
            '\u{FF01}'..='\u{FF5E}' => char::from_u32((c as u32) - 0xFEE0).unwrap_or(c),
            _ => c,
        })
        .collect()
}

pub(super) fn dedup_sorted(mut values: Vec<String>) -> Vec<String> {
    values.sort();
    values.dedup();
    values
}

pub(super) fn dedup_preserve(values: Vec<String>) -> Vec<String> {
    let mut seen = HashSet::new();
    let mut out = Vec::new();
    for v in values {
        if seen.insert(v.clone()) {
            out.push(v);
        }
    }
    out
}

File: web/src/core/name-utils.ts
export function foldFullwidthAscii(input: string): string {
  let out = "";
  for (const char of input) {
    const code = char.codePointAt(0) ?? 0;
    if (code === 0x3000) {
      out += " ";
    } else if (code >= 0xff01 && code <= 0xff5e) {
      out += String.fromCodePoint(code - 0xfee0);
    } else {
      out += char;
    }
  }
  return out;
}

export function normalizeName(value: string): string {
  return foldFullwidthAscii(value)
    .toLowerCase()
    .replace(/\s+/g, "")
    .replace(/['“”]/g, '"');
}

export function dedupPreserve(values: string[]): string[] {
  const out: string[] = [];
  const seen = new Set<string>();
  for (const value of values) {
    if (!seen.has(value)) {
      seen.add(value);
      out.push(value);
    }
  }
  return out;
}

export function dedupSorted(values: string[]): string[] {
  return [...new Set(values)].sort();
}
